// testdouble/lib/store/calls.js
// Generated by CoffeeScript 1.10.0
(function() {
  var _, argsMatch, lastCall, store;

  _ = require('lodash');

  store = require('./index');

  argsMatch = require('./../args-match');

  lastCall = null;

  store.onReset(function() {
    return lastCall = null;
  });

  module.exports = {
    log: function(testDouble, args, context) {
      store["for"](testDouble).calls.push({
        args: args,
        context: context
      });
      return lastCall = {
        testDouble: testDouble,
        args: args,
        context: context
      };
    },
    pop: function() {
      var call;
      return _.tap((call = lastCall), function(call) {
        lastCall = null;
        if (call != null) {
          return store["for"](call.testDouble).calls.pop();
        }
      });
    },
    wasInvoked: function(testDouble, args, config) {
      var matchingInvocationCount;
      matchingInvocationCount = this.where(testDouble, args, config).length;
      if (config.times != null) {
        return matchingInvocationCount === config.times;
      } else {
        return matchingInvocationCount > 0;
      }
    },
    where: function(testDouble, args, config) {
      return _.select(store["for"](testDouble).calls, function(call) {
        return argsMatch(args, call.args, config);
      });
    },
    "for": function(testDouble) {
      return store["for"](testDouble).calls;
    }
  };

}).call(this);
